lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2197,6 +2197,21 @@ bool TypeChecker::typeCheckCondition(Expr *&expr, DeclContext *dc) {
     virtual Expr *appliedSolution(constraints::Solution &solution,
                                   Expr *expr) {
       auto &cs = solution.getConstraintSystem();
+
+      // Since conversion constraint is going to enforce actual coercion
+      // only in some cases e.g. function application, we need to explicitly
+      // coerce expression to boolean type here before proceeding.
+      if (!expr->getType()->isBool()) {
```

Comment:
I don't quite follow the comment. Why is the conversion constraint only enforcing coercion for function applications?